Documentation/umd cli (#283)
CoherentLabs/GameUIComponents: docs/content/en/command-line-interface/cli.md

The coherent-guic-cli is a tool that helps you create and develop components from the GameUIComponents suite. It can also setup a development environment with live reload that will allow you to easily create a UI using the components.

# Getting Started

To install the cli open a console and type:

```
npm i -g coherent-guic-cli
```

This will install the cli as a global module and will make it available from everywhere.

## Available Commands

|Command   |Description   |Arguments   |Usage   |
|---|---|---|---|
|create-project               |Create a new project from template                 | `<template>` `<directory>`|`coherent-guic-cli create-project Webpack ./my-folder`|
|create | [name], [directory] |           Create a new component | `coherent-guic-cli create my-component ./my-folder`|
|build |[--watch]        | start the server|`coherent-guic-cli build`|
|start:demo| N/A|              Start a development server and host the demo|`coherent-guic-cli start:demo`|
|build:demo| [--dev] |      Create a production bundle of the demo.

## Usage

> ### Setting up an Environment

Reviewer: This section is a bit strange. You are giving an example of how to create a project with webpack and then an example is given with vite. The commands are the same for both - `npm run dev`.
Either combine both into a single example and usage or split both sections with headings such as
```
#### Project with webpack

...

#### Project with Vite

...
```
for example.

Author: Added clarification for vite and webpack. The command is the same - yes, it was on purpose.

Reviewer 2: Better! 🙇

If you want to create a UI using the components - use the `create-project` command. It allows you to setup a project that uses [Webpack](https://webpack.js.org/) or [Vite](https://vitejs.dev/). Open a console and type:

> ```

Author: I don't see a warning and the spellcheck passes (or doesn't run at all, idk) - added a new line.

coherent-guic-cli create-project Webpack ./my-folder
```

This will automatically install all dependencies. After the project is ready - navigate to `my-folder` and type
```
npm run dev
```

this will start a development server with live reload enabled.

`<template>` can be either 'webpack' or 'vite'.

Example:
```
coherent-guic-cli create-project vite ./my-folder
```

This will create a new project inside `./my-folder` and will install the dependencies. After that you can navigate to the folder and start a development server:
```
cd ./my-folder
npm run dev
```

### Creating a Component

> After the installation is complete open a console and run:

Reviewer: Installation of what exactly - the cli? Run, where? :D my-folder? :D We are a bit confused here because you are giving an example above of creating a project and maybe it should continue here with creating a new component to the new project but the context of this topic is different.

Author: Yep, it used to make some kind of sense before the changes, but now that I added the docs about create-project it is confusing. Added clarification about what this section is about - check it and let me know if it's still confusing.

`coherent-guic-cli create my-first-guic components`

The create command accepts two positional arguments - the name of the component and its containing folder. The name of the component will be used for the registration of the custom HTML element and for the name of the component class. So the above command will generate the following files:

```
\components:
|
\---gameface-my-first-guic
    |   coherent-gameface-components-theme.css
    |   index.js
    |   package.json
    |   README.md
    |   script.js
    |   style.css
    |   template.html
    |
    \--demo
            demo.html
            demo.js
```

> 
> The registered custom element will look like this:

Reviewer: ?
```suggestion
The registered custom element can be used like this:
```

```html
<gameface-my-first-guic class="my-first-guic-component">
    <component-slot data-name="name">my-first-guic</component-slot>
</gameface-my-first-guic>
```

And the custom element's definition looks like this:

```js
class GamefaceMyFirstGuic extends BaseComponent {
   constructor() {
        super();
        this.template = template;
        this.init = this.init.bind(this);
    }

    init(data) {
        this.setupTemplate(data, () => {
            components.renderOnce(this);
            // attach event handlers here
        });
    }

    connectedCallback() {
        components.loadResource(this)
            .then(this.init)
            .catch(err => console.error(err));
    }
}
```

> Note that the 'gameface' prefix is added to the component name and to the containing folder.

Reviewer: ```suggestion
Note that the `gameface` prefix is added to the component name and to the containing folder.
```

To build the component run:

1. `npm i` in the component's folder to install the dependencies.
2. `coherent-guic-cli build` to bundle the source of the component
3. `coherent-guic-cli build:demo` and load the demo.html folder or
4. `coherent-guic-cli start:demo` to start a development server that will host the demo page and watch for changes

## Commands:

* coherent-guic-cli create <name> <destination> - Creates a new component with
given name at a given destination folder.
* coherent-guic-cli build - Creates cjs and umd bundles of the component. The
bundles are created in a cjs/ and umd/ folders next to the component source.
* coherent-guic-cli build --watch - Pass --watch to watch for changes to the source
and rebuild.
* coherent-guic-cli build:demo - Create a production build of the demo that bundles
* coherent-guic-cli build:demo --dev - Pass the --dev option to create a development non-minified build

> * coherent-guic-cli start:demo - Start a development server which hosts the demo

Reviewer: Don't we duplicate here the table with the `Available commands` above?

Author: Yes, we do.... removed :D
